mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+		}
+	}
+	return isPk
+}
+
+// getTypeModsAndID returns ID and mods of column datatype
+func getTypeModsAndID(columnType string) (string, []int64) {
+	// There are no methods in pincap parser to retirieve ID and mods
+	// We will process columnType eg:'varchar(40)' and split ID from the string.
+	// We retrieve mods using regex expression and convert it to INT64
+	id := columnType
+	var mods []int64
+	if strings.Contains(columnType, "(") {
+		id = strings.Split(columnType, "(")[0]
+		// For 'set' and 'enum' datatypes, values provided are not maxLength
+		if id == "set" || id == "enum" {
+			return id, nil
+		}
+		values := valuesRegexp.FindString(columnType)
+		strMods := strings.Split(values[1:len(values)-1], ",")
+		for _, i := range strMods {
+			j, err := strconv.ParseInt(i, 10, 64)
+			if err != nil {
+				panic(err)
+			}
+			mods = append(mods, j)
+		}
+	}
+	// 'BINARY' keyword will be added to all blob datatypes by parser.
+	// Eg: mediumblob BINARY. It needs to be ignored to retrieve ID
+	if strings.Contains(id, " ") {
+		id = strings.Split(columnType, " ")[0]
+	}
+	return id, mods
+}
+
+// handleParseError handles error while parsing mysqldump
+// statements and attempts at creating parsable query.
+// Error can be due to insert statement or unsupported Spatial
+// datatypes in create statement.
+func handleParseError(conv *internal.Conv, query string, err error, l [][]byte) ([]ast.StmtNode, bool) {
+	// Check if error is due to Insert statement.
+	insertStmt := insertRegexp.FindString(query)
+	if insertStmt != "" {
+		// Likely causes of failing to parse Insert statement:
+		// a) Due to some invalid value
+		// b) Query size is more than what pingcap parser could handle (more than 40MB in size)
+		return handleInsertStatement(conv, query, insertStmt)
+	}
+
+	// Check if error is due to spatial datatype as it is not supported by Pingcap parser.
+	var isSpatial bool
+	for _, spatial := range MysqlSpatialDataTypes {
+		if strings.Contains(strings.ToLower(err.Error()), spatial) && isSpatial == false {
+			isSpatial = true
+			if conv.SchemaMode() {
+				conv.Unexpected(fmt.Sprintf("Unsupported datatype '%s' encountered while parsing following statement at line number %d : \n%s", spatial, len(l), query))
+				internal.VerbosePrintf("Converting datatype '%s' to 'Text' and retrying to parse the statement\n", spatial)
+			}
+		}
+	}
+	if isSpatial {
+		return handleSpatialDatatype(conv, query, l)
+	}
+
+	return nil, true
+}
+
+// handleInsertStatement handles error in parsing the insert statement.
+// We deal with this case by extracting all rows and creating
+// extended insert statements. Then we parse one Insert statement
+// at a time, ensuring no size issue and skipping only invalid entries.
+func handleInsertStatement(conv *internal.Conv, query, insertStmt string) ([]ast.StmtNode, bool) {
+	var stmts []ast.StmtNode
+	values := valuesRegexp.FindAllString(query, -1)
```

Comment:
Just checking on the scalability issues here.
If I have a really huge mysql table, say 100GB in one table, would this show up as one big INSERT statement? Which would mean we'd have to have all of the 100GB in memory to do this regexp parse?

I recall chatting about this earlier, but I can't recall the outcome -- sorry!
It would be worth adding a comment on this here.

Reply:
- Firstly, it cannot be 100 GB in a single insert statement. In `mysqldump` there is an option to set size of insert statement block, default being 24 MB. Thus, insert statements will be broken down in multiple insert statement of size 24 MB each. 

`--max-allowed-packet=value

The maximum size of the buffer for client/server communication. The default is 24MB, the maximum is 1GB.` ([LINK](https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_max-allowed-packet))

-  We have tested that pingcap parser is able to handle chunk upto 40 MB. If for some reason, user sets max size of chunk to 64 MB while converting to mysqldump, then `handleInsertStatement()` would be able to handle that scenario.

Reply:
Yes, thanks for the refresher.
How about we add a comment to readAndParseChunk to explain this -- I think it is worth documenting (at least for my sake!).
I added a separate common with a draft comment.